src/library/scala/collection/Iterable.scala

```diff
@@ -549,21 +549,22 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
     *
     */
   def groupBy[K](f: A => K): immutable.Map[K, C] = {
-    val m = mutable.Map.empty[K, Builder[A, C]]
+    val m = mutable.HashMap.empty[K, Any /*C | Builder[A, C] */]
     val it = iterator
-    while (it.hasNext) {
-      val elem = it.next()
-      val key = f(elem)
-      val bldr = m.getOrElseUpdate(key, newSpecificBuilder)
-      bldr += elem
-    }
-    var result = immutable.HashMap.empty[K, C]
-    val mapIt = m.iterator
-    while (mapIt.hasNext) {
-      val (k, v) = mapIt.next()
-      result = result.updated(k, v.result())
+    if (it.isEmpty) immutable.Map.empty
+    else {
+      while (it.hasNext) {
+        val elem = it.next()
+        val key  = f(elem)
+        val bldr = m.getOrElseUpdate(key, newSpecificBuilder).asInstanceOf[Builder[A, C]]
```

Comment:
could we allow a hook here so that, for some collection to avoid the builder, where the receiver can cope with C | Builder[C] | some other type?
Also I see a common pattern that the groupBy is very distinct
e.g. `mySet.groupBy(_.id)`
both of these cases we create builders that we could optimise away.

Reply:
I don't understand this suggestion. You could provide concrete examples?

Note that the 2.13.x collections API has `groupMapReduce` which is use useful when you want just a single element as the value in each map entry.

```
scala> (-100 to 100).groupMapReduce(x => x % 5)(x => math.abs(x))(_ + _)
val res4: scala.collection.immutable.Map[Int,Int] = HashMap(0 -> 2100, -3 -> 1010, 1 -> 970, -4 -> 1030, 2 -> 990, 3 -> 1010, -1 -> 970, 4 -> 1030, -2 -> 990)
```

Reply:
Ah - didn't know of that API. Its a more general solution